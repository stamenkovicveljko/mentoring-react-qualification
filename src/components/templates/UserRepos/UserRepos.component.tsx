import React, { useState } from "react";
import IGithubUser from "../../../models/GithubUser";
import IGithubRepo from "../../../models/GithubRepo";
import ProfileInfo from "../../organisms/ProfileInfo/ProfileInfo.component";
import Input from "../../atoms/Input/Input.component";
import List from "../../molecules/List/List.component";
import Link from "../../atoms/Link/Link.component";
import * as Styled from "./UserRepos.styles";

interface Props {
  user: IGithubUser;
  repos: IGithubRepo[];
}

const UserRepos = ({ user, repos }: Props) => {
  const [term, setTerm] = useState<string>("");

  const handleTermOnChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    setTerm(ev.target.value);
  };

  return (
    <Styled.Container>
      <ProfileInfo user={user} className="profile-info" />
      <div>
        <Input
          value={term}
          onChange={handleTermOnChange}
          placeholder="Search repos"
        />
        <List
          items={repos
            .filter(({ name }) =>
              name.toLowerCase().includes(term.toLowerCase())
            )
            .map((repo) => (
              <Link to={repo.html_url}>{repo.name}</Link>
            ))}
        />
      </div>
    </Styled.Container>
  );
};
export default UserRepos;
