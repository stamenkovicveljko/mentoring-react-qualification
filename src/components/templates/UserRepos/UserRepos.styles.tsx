import styled from "styled-components";

export const Container = styled.div`
  width: 100%;

  .profile-info + * {
    margin-top: 64px;
  }
`;
