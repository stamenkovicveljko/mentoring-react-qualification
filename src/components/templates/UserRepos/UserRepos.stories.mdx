import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import UserRepos from "./UserRepos.component";

<Meta title="UserRepos" component={UserRepos} />

# UserRepos Component

Creates searchable list of specified user's repos from the github.

## Implementation

It relies on `user` and `repos` to be defined.

_User model_

```ts
interface IGithubUser {
  name: string;
  login: string;
  avatar_url: string;
  html_url: string;
}
```

_Repo model_

```ts
interface IGithubRepo {
  name: string;
  html_url: string;
}
```

<Preview>
  <Story name="default">
    <UserRepos
      user={{
        name: "Some user",
        login: "storiez",
        avatar_url:
          "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
        html_url:
          "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
      }}
      repos={[
        {
          name: "some repo 1",
          html_url: "https://google.com",
        },
        {
          name: "some repo 2",
          html_url: "https://google.com",
        },
        {
          name: "some repo 3",
          html_url: "https://google.com",
        },
      ]}
    />
  </Story>
</Preview>

## Props

<Props of={UserRepos} />
