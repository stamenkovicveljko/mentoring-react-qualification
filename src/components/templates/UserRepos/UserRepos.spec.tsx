import React from "react";
import renderer from "react-test-renderer";
import UserRepos from "./UserRepos.component";
import { mount, ReactWrapper } from "enzyme";

let wrapper: ReactWrapper;

describe("UserRepos", () => {
  it("renders as expected", () => {
    const component = renderer.create(
      <UserRepos
        user={{
          name: "Some user",
          login: "storiez",
          avatar_url:
            "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
          html_url:
            "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
        }}
        repos={[
          {
            name: "some repo 1",
            html_url: "https://google.com",
          },
          {
            name: "some repo 2",
            html_url: "https://google.com",
          },
          {
            name: "some repo 3",
            html_url: "https://google.com",
          },
        ]}
      />
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe("given that repos are provided", () => {
    it("filters repoes based on user input", () => {
      wrapper = mount(
        <UserRepos
          user={{
            name: "Some user",
            login: "storiez",
            avatar_url:
              "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
            html_url:
              "https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg",
          }}
          repos={[
            {
              name: "some repo 1",
              html_url: "https://google.com",
            },
            {
              name: "some repo 2",
              html_url: "https://google.com",
            },
            {
              name: "some repo 3",
              html_url: "https://google.com",
            },
          ]}
        />
      );

      wrapper.find("input").simulate("change", {
        target: {
          value: "2",
        },
      });

      expect(wrapper.find("li").length).toBe(1);
    });
  });
});
