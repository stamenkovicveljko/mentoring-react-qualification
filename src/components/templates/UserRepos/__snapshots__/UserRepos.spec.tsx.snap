// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepos renders as expected 1`] = `
<div
  className="sc-AxhUy kgzkpi"
>
  <div
    className="sc-AxjAm whBCV profile-info"
  >
    <img
      alt="Profile img"
      src="https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg"
    />
    <div>
      <div>
        <span>
          Username
        </span>
        <span
          className="user-data"
        >
          storiez
        </span>
      </div>
      <div>
        <span>
          Name
        </span>
        <span
          className="user-data"
        >
          Some user
        </span>
      </div>
      <div>
        <span>
          Github Profile:
        </span>
        <span
          className="user-data"
        >
          <a
            className="sc-AxirZ ldeUjV"
            href="https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg"
          >
            https://velikirecnik.com/wp-content/uploads/2017/11/Avatar.jpeg
          </a>
        </span>
      </div>
    </div>
  </div>
  <div>
    <div
      className="sc-AxiKw jOphat"
    >
      <input
        onChange={[Function]}
        placeholder="Search repos"
        value=""
      />
    </div>
    <ul
      className="sc-AxhCb ivQGmQ"
    >
      <li>
        <a
          className="sc-AxirZ ldeUjV"
          href="https://google.com"
        >
          some repo 1
        </a>
      </li>
      <li>
        <a
          className="sc-AxirZ ldeUjV"
          href="https://google.com"
        >
          some repo 2
        </a>
      </li>
      <li>
        <a
          className="sc-AxirZ ldeUjV"
          href="https://google.com"
        >
          some repo 3
        </a>
      </li>
    </ul>
  </div>
</div>
`;
